### DISCUSSION TITLE
Magit selections with shift-selection-mode on and transient-mark-mode off

### DISCUSSION THREAD

**Discussion by farblos** [magit/magit#5110] [Support] [answered]
### Magit

```
Magit 20240311.1442 [>= 3.3.0.50-git], Transient 0.5.3, Git 2.43.0, Emacs 29.2, gnu/linux
```

### Emacs

```
GNU Emacs 29.2 (build 2, x86_64-pc-linux-gnu, X toolkit, cairo version 1.18.0, Xaw3d scroll bars) of 2024-02-27, modified by Debian
```

### Description

Magit selections (in staged or unstaged diffs, for example) behave somewhat surpirsingly when `shift-selection-mode` is on and `transient-mark-mode` off.  From a [previous discussion with Eli](https://lists.gnu.org/archive/html/bug-gnu-emacs/2023-10/msg01836.html) I understood that this is a slightly undersupported feature in Emacs in general, so I'd like to double-check here before opening an issue.

### Reproducer

- Start `emacs -Q`, `M-: (package-activate-all) RET`
- `C-u C-x g <some-Git-controlled-directory> RET`
- Visit some staged or unstaged changes, press `S-<down>` on some diff:
  
  Point does not move (expected), Magit selection is displayed (expected)

- `C-g` (to get rid of the active region)

- `M-x transient-mark-mode RET`

- Press `S-<down>` on some diff:

  Message `Mark set` (expected), point does not move (expected),
  Magit selection is **not** displayed (**not** expected)

Only on second `S-<down>` does the Magit selection show up.

### Educated Guess

I think that at the point in time when the Magit selection is handled (`post-command-hook`)
the shift-selection code has not yet set up this temporary transient mark mode, so that
`region-active-p` still returns `nil`.

I can try to follow up, but please let me know first if there is a general interest in getting this
fixed...

**Comment by tarsius** [comment, marked as answer]
Magit uses specialized `previous-/next-line` commands to deal with shift-selection. What those do might not be the best they could do. First see the docstring of `magit-next-line`.

> Message Mark set (expected), point does not move (expected),

Why is the latter expected by you? The first `next-line` *does* move to the next line, even when using shift-selection. The magit commands *prevent* that because the region works a bit differently as explained in the mentioned docstrings.

As a quick check I replaced `(push-mark nil nil t)` with `(handle-shift-selection)` in these functions. Now sometimes it works, as *Magit* expects it: point does not move on the first invocation and the line is highlighted. But sometimes it does not: the line is also highlighted, but point moves.

Maybe, after some reading and with the help of some debug statements, I can find a way to make it always work as intended. I am very very busy right now, so I won't do that soon.

As a workaround I believe you can move down an additional line and then up again; then the single, selected line is highlighted as expected.

**Reply by farblos** [reply]
> I am very very busy right now, so I won't do that soon.

That's fine, thanks for the quick reply.

At least I understood that there *is* at least some general interest in handling this case, and so I also might try to investigate here ...
